src/main.rs

```
@@ -1,49 +1,171 @@
+    let config = load_config(config_path).context("failed to load config")?;
 
     match config.mode {
         Mode::Client(client_config) => {
-            let client = Client::try_new(client_config, config.tls)?;
-            let stop_sender = client.stop_sender();
-            ctrlc::set_handler(move || {
-                if let Err(err) = stop_sender.send(true) {
-                    error!("could not stop: {err}");
-                }
-            })
-            .context("could not set Ctrl-C handler")?;
-            runtime.block_on(client.run())
-        }
-        Mode::Server(server_config) => runtime.block_on(async move {
-            Server::try_new(server_config, config.tls)
-                .map(|server| server.run())?
-                .await
-        }),
+            Client::try_new(client_config, config.tls).context("failed to build client")
+        }
+        _ => Err(anyhow::anyhow!("config is not in client mode")),
     }
 }
+
+fn browse_handler(app_weak: slint::Weak<AppWindow>) {
+    TOKIO_RUNTIME.spawn(async move {
+        let file_path = rfd::FileDialog::new()
+            .pick_file()
+            .map(|p| p.display().to_string());
+
+        let _ = slint::invoke_from_event_loop(move || match app_weak.upgrade() {
+            Some(app) => {
+                if let Some(p) = file_path {
+                    app.set_from_file(LineEditInfo {
+                        text: SharedString::from(p),
+                        ..app.get_from_file()
+                    })
+                }
+                app.set_freeze(false);
+            }
+            None => {
+                error!("app downgraded before setting selected file");
+            }
+        });
+    });
+}
+
+fn connect_handler(app_weak: slint::Weak<AppWindow>, client: Client) {
+    TOKIO_RUNTIME.spawn(async move {
+        if let Err(e) = client.run().await {
+            let _ = slint::invoke_from_event_loop(move || match app_weak.upgrade() {
```

silently ignoring possible error